Cancer-GEMINI
=============================================================================

Overview
========
Cancer-GEMINI is an adaptation of GEMINI intended for the improved identification of
biologically and clincally relevant tumor variants from multi-sample and longitudinal
tumor sequencing data. Using a GEMINI-compatible database (generated from an annotated 
VCF file), Cancer-GEMINI is able to filter tumor variants based on included genomic
annotations and various allele frequency signatures. 


Installation
============


Documentation
================
The official documentation is here:

Since Cancer-GEMINI retains much of the functionality of GEMINI, it may also be 
helpful to refer to GEMINI's official documentation which can be found [here](http://gemini.readthedocs.org/en/latest/).


Citation
================
If you use Cancer-GEMINI in your research, please cite the following manuscript:


Acknowledgements
================
Cancer-GEMINI is being developed in the Quinlan lab (quinlanlab.org) at the University
of Utah and is led by Tom Nicholas and Aaron Quinlan.  Substantial contributions and discussions 
have been made by Brent Pedersen, Yi Qiao, Xiaomeng Huang, and Gabor Marth.
